README.md
[![Waffle.io - Columns and their card count](https://badge.waffle.io/adsabs/object_service.svg?columns=all)](https://waffle.io/adsabs/object_service)
[![Build Status](https://travis-ci.org/adsabs/object_service.svg?branch=master)](https://travis-ci.org/adsabs/object_service)
[![Coverage Status](https://coveralls.io/repos/github/adsabs/object_service/badge.svg?branch=master)](https://coveralls.io/github/adsabs/object_service?branch=master)
[![Code Climate](https://codeclimate.com/github/adsabs/object_service/badges/gpa.svg)](https://codeclimate.com/github/adsabs/object_service)
[![Issue Count](https://codeclimate.com/github/adsabs/object_service/badges/issue_count.svg)](https://codeclimate.com/github/adsabs/object_service)

Micro service implementation of ADS object searches. This micro service supports search of the literature for bibliographic records relevant to one or more astronomical objects or for a specified position on the sky. An object name query searches used the SIMBAD and/or NED services to find relevant records in the ADS.

## SIMBAD implementation

The following queries are supported to retrieve information from SIMBAD:

#### Object Search
This search accepts either a list of object names or a list of SIMBAD object identifiers (POST queries). Both queries return JSON where the keys are object names and the values are dictionaries containing the (SIMBAD) canonical object name and the SIMBAD identifier. For example

    {
		"LMC" : {"canonical": "LMC",
		         "id": 12345
				 },
		"Andromeda": {"canonical": "ANDROMEDA",
		              "id": 342536
	             },
		"FooBar": None
	}

where a value of `None` means that the associated object name could not be identified in the SIMBAD database.

1. Retrieve SIMBAD identifiers for a string of object names

Using the Python "requests" module, this works as follows:

    headers = {'Content-type': 'application/json', 'Accept':'text/plain'}
	queryURL = 'http://localhost:4000/'
	payload = {"objects": ["M11", "M13", "Andromeda"]}
	r = requests.post(queryURL, data=json.dumps(payload), headers=headers)

and the results, returned through `r.json()`, are

    {"M11":{"canonical":"M  11", "id":2613692}, "M13":{"canonical": "M  13", "id":2894585}, "Andromeda":{"canonical":"ANDROMEDA", "id":1575544}}

2. Retrieve canonical object names for a list of SIMBAD identifiers

In this case this works as follows:

    headers = {'Content-type': 'application/json', 'Accept':'text/plain'}
	queryURL = 'http://localhost:4000/'
	payload = {"identifiers": [3133169, 1575544, 2419335, 3253618]}
	r = requests.post(queryURL, data=json.dumps(payload), headers=headers)

and the results in this case are

    {"3253618":{"canonical": "SMC", "id":3253618}, "1575544":{"canonical": "M  31", "id":1575544}, "2419335":{"canonical": "GAL CENTER", "id":2419335}, "3133169":{"canonical": "LMC", "id": 3133169}}

#### Positional ("Cone") Search
This search takes a position string and (optionally) a search radius. Position searches locate the papers dealing with celestial objects located within the specified radius of the specified position. The syntax for position searches is: 

    RA ±Dec : radius 

where RA and Dec are right ascention and declination J2000 positions, expressed in decimal degrees or in sexagesimal notation (hours minutes seconds and degrees arcmin and arcsec). The plus or minus sign before the declination is mandatory. The search radius may be given in decimal or sexagesimal degrees. For example, a 10' radius may be written as 0.1667 or 0 10 . The default search radius is 2' (0 2 = 0.033 deg). Some examples of allowed position strings:

    05 23 34.6 -69 45 22:0 6 (or 05h23m34.6s -69d45m22s:0m6s)
	05 23 34.6 -69 45 22:0.166666 (or 05h23m34.6s -69d45m22s:0.166666)
	80.89416667 -69.75611111:0.166666

The query is performed as follows

	pos_string = "80.89416667 -69.75611111:0.166666"
	queryURL = 'http://localhost:4000/pos/%s' % pos_string
	r = requests.get(queryURL)

This search returns the following results:

    {u'data': [u'1998A&A...335L..65H', u'2012yCat.1322....0Z', u'2008AcA....58..293S', u'2010AcA....60....1P', u'1974A&AS...18...47B', u'2011ApJS..197...16W', u'2009ApJS..184..172G', u'1972PASP...84..365H', u'1964IrAJ....6..241A', u'1999A&AS..139..277H', u'2003yCat.2246....0C', u'1988MNRAS.230..215B', u'2007PASP..119...19S', u'2010AcA....60..179P', u'1999AcA....49..521P', u'2011AcA....61..199P', u'2000A&A...355L..27H', u'2008AcA....58..163S', u'2012ApJ...747..107K', u'2008ApJS..178...56F', u'2011AJ....142...48W', u'2003A&A...405..111G', u'2008AJ....136...18W', u'2003yCat.1289....0Z', u'2000A&AS..145...11M', u'2003AcA....53...93S', u'2007AJ....134.1963F', u'2012ApJ...755...40P', u'2002AJ....124.2039K', u'2002LEDA.........0P', u'1976A&AS...24...35S', u'2011A&A...536A..60S', u'2000A&AS..143..391S', u'2009AcA....59....1S', u'1999AJ....117..238B', u'2002AcA....52..143S', u'1963IrAJ....6..127L', u'2006MNRAS.373..521R', u'2007ApJ...663..249D', u'2009AcA....59..239S', u'1988IRASP.C......0J', u'1988NASAR1190....1B', u'2001A&A...377..945C', u'2011AcA....61..103G', u'2008MNRAS.389..678B', u'2002AJ....124.3241S', u'2010A&A...514A...1I']}


CHANGES.md
### 1.0.66
* Removed service token, the service will use user's credentials

### 1.0.65

* Fix bug when there is a caret in the query string

### 1.0.64

* Better error handling for failed cone search

### 1.0.63

* Cone search fix and restriction of allowed coordinate formats

### 1.0.62

* Switch to https for all queries to NED and SIMBAD

### 1.0.61

* Switched off synonym replacement for ADS metadata search

### 1.0.60

* Improved coverage of unittests

### 1.0.59

* Update of unittests

### 1.0.58

* Fix for failing results verification in cone search

### 1.0.57

* Updated adsmutils requirement
* Removed some unnecessary requirements

### 1.0.56

* Updated adsmutils requirement
* Removed some unnecessary requirements

### 1.0.55

* consistent use of current_app.client 
* removed client.py

### 1.0.54

* reinstatement of timeouts
* use current_app client in utils.py

### 1.0.53

* temporary removal of timeouts in POST requests

### 1.0.52

* Better coordinate handling and transforming

### 1.0.51

* Fixed bug in query parameters NED cone search

### 1.0.50

* Improved log messages for exceptions/errors
* Check SIMBAD identifiers supplied for validity

### 1.0.49
 
* Updated adsmutils requirement

### 1.0.48
 
* Enabled JSON stdout logging and HTTP connection pool

### 1.0.47

* bugfix: Github issue 78 (trailing spaces in object names)

### 1.0.46

* bugfix in returned data processing

### 1.0.45

* Update of HTTP status codes for nedsrv errors

### 1.0.44

* For cone search, limit the potential number of objects returned

### 1.0.43

* Unittest coverage improvement
* Cone search support

### 1.0.42

* Maintenance release (bug in error message)

### 1.0.41

* Maintenance release (improved unittest coverage)

### 1.0.40

* Some minor bug fixes to deal with pathological erroneous queries

### 1.0.39

* Implementation of new query parsing method using the luqum (LUcene QUery Manipolator) module
  * luqum credentials: developed by Jurismarchés (https://github.com/jurismarches/luqum)
  * License: GNU General Public License v3.0
  * the module has been adapted to be compatible with Python 2.7
* The luqum module is used to create a query tree from the Solr query string to identify the object: modifiers
* This fixes the bug that e.g. "object:Andromeda OR abs:Andromeda" gets translated into "simbid:123 OR 0:123"
* Enhancement: an object: query is now translated into an OR of simbid:, nedid: and abs: queries, with
  database:astronomy as restriction

### 1.0.38

* Fix of incorrect urls in log messages

### 1.0.37

* Nedsrv endpoint: provide info when 500 is thrown

### 1.0.36

* Unittest coverage improvement and code cleanup

### 1.0.35

* Integration of NED into object search

### 1.0.34

* Bugfix for nedsrv endpoint

### 1.0.33

* Endpoint 'classic' renamed to 'nedsrv'

### 1.0.32

* Fix of mime-type for ASCII output. Add filtering by month.

### 1.0.31

* additional output format and filters for "Classic" endpoint

### 1.0.30

* remove dependency on astroquery

### 1.0.29

* endpoint to support ADS Classic NED query

### 1.0.28

* update of query sent to SIMBAD TAP (no uppercasing)

### 1.0.27

* specify version for adsmutils in requirements.txt

### 1.0.26

* Special User Agent string for external requests

### 1.0.25

* No more local caching 

### 1.0.24

* More code cleanup

### 1.0.23

* Code cleanup and waffle implementation

### 1.0.22

* Undo the replacement of spaces with underscores in identifiers

### 1.0.21

* App instantiation through ADSFlask (ADSMicroserviceUtils)

### 1.0.20

* Unittesting using py.test
* Renamed app directory from 'service' to 'object_service'
* Separate dev-requirements.txt
* Removed unused client.py
* Code cleanup

### 1.0.19

* Implementation of data retrieval to support NED facet

### 1.0.18

* Update requirements.txt (versioning) + cleanup

### 1.0.17

* fix of Github issue 38: catch ReadTimeout exceptions

### 1.0.16

* logging of URL TAP service used

### 1.0.15

* more detailed logging

### 1.0.14

* timeout exception implemented for SIMBAD requests

### 1.0.13

* small update

### 1.0.12

* Update of logfile name, making use of ENVIRONMENT variable

### 1.0.11

* Fixed bug Github29 (allowing for more complex nested operator queries)

### 1.0.10

* Fixed bug Github27

### 1.0.9

* Fixed bug Github24

### 1.0.8

* deal with format of BBB requests in case of query input

### 1.0.7

* cleanup and allow query as input

### 1.0.6

* for identifier query, output dictionary has identifiers as keys

### 1.0.5

* process top level facets properly

### 1.0.4

* addition of facets as allowed input type (Github 15)

### 1.0.3

* bug fix (Github issue 10) and cleanup of SIMBAD module

### 1.0.2

* refactor of SIMBAD object search to facilitate integration in front-end

### 1.0.1

* unittests for SIMBAD services

### 1.0.0

* initial version
* only support for SIMBAD object queries


service/views.py
from flask import current_app, request
from flask_restful import Resource
from flask_discoverer import advertise
from flask_cache import Cache
from SIMBAD import get_simbad_data
from SIMBAD import do_position_query
from SIMBAD import parse_position_string
from NED import get_ned_data
from utils import get_objects_from_query_string
import time
import timeout_decorator

class IncorrectPositionFormatError(Exception):
    pass

class ObjectSearch(Resource):

    """Return object identifiers for a given object string"""
    scopes = []
    rate_limit = [1000, 60 * 60 * 24]
    decorators = [advertise('scopes', 'rate_limit')]

    def post(self):
        stime = time.time()
        # Get the supplied list of identifiers
        identifiers = []
        objects = []
        facets = []
        input_type = None
        # determine whether a source for the data was specified
        try:
            source = request.json['source'].lower()
        except:
            source = 'simbad'
        for itype in ['identifiers', 'objects', 'facets']:
            try:
                identifiers = request.json[itype]
                identifiers = map(str, identifiers)
                input_type  = itype
            except:
                pass
        if not input_type:
            current_app.logger.error('No identifiers and objects were specified for SIMBAD object query')
            return {"Error": "Unable to get results!",
                    "Error Info": "No identifiers/objects found in POST body"}, 200
        # We should either have a list of identifiers, a list of object names or a list of facets
        if len(identifiers) == 0 and len(objects) == 0 and len(facets) == 0:
            current_app.logger.error('No identifiers, objects or facets were specified for SIMBAD object query')
            return {"Error": "Unable to get results!",
                    "Error Info": "No identifiers/objects found in POST body"}, 200
        # How many iden identifiers do we have?
        id_num = len(identifiers)
        if id_num == 0:
            return {"Error": "Unable to get results!",
                    "Error Info": "No identifiers/objects found in POST body"}, 200
        # Now check if we have anything cached for them
        cached = {id:current_app.cache.get(id.upper()) for id in identifiers if current_app.cache.get(id.upper())}
        if source in ['simbad','ned'] and len(identifiers) > 0:
             # If we have cached values, filter those out from the initial list
            if cached:
                current_app.logger.debug('Received %s %s. Using %s entries from cache.' % (id_num, input_type, len(cached)))
                identifiers = [id for id in identifiers if not current_app.cache.get(id.upper())]
            if identifiers:
                # We have identifiers, not found in the cache
                if source == 'simbad':
                    result = get_simbad_data(identifiers, input_type)
                else:
                    result = get_ned_data(identifiers, input_type)
                if 'Error' in result:
                    # An error was returned!
                    err_msg = result['Error Info']
                    current_app.logger.error('Failed to find data for %s %s query (%s)!'%(source.upper(), input_type,err_msg))
                    return result
                else:
                    # We have results!
                    duration = time.time() - stime
                    current_app.logger.info('Found objects for %s %s in %s user seconds.' % (source.upper(), input_type, duration))
                    # Before returning results, cache them
                    for ident, value in result['data'].items():
                        current_app.cache.set(ident.upper(), value, timeout=current_app.config.get('OBJECTS_CACHE_TIMEOUT'))
                    # Now pick the entries in the results that correspond with the original object names
                    if input_type == 'objects':
                        result['data'] = {k: result['data'].get(k.upper()) for k in identifiers}
                    # If we had results from cache, merge these in
                    if cached:
                        res = cached.copy()
                        res.update(result.get('data',{}))
                        return res
                    # Otherwise just send back the results
                    else:
                        return result.get('data',{})
            elif cached:
                # We only had cached results
                return cached
            else:
                # This should never happen
                current_app.logger.error('No data found, even though we had %s! Should never happen!'%input_type)
                result = {
                    "Error": "Failed to find data for %s %s query!"%(source.upper(), input_type),
                    "Error Info": "No results found, where results were expected! Needs attention!"
                    }
                return result

class PositionSearch(Resource):

    """Return publication information for a cone search"""
    scopes = []
    rate_limit = [1000, 60 * 60 * 24]
    decorators = [advertise('scopes', 'rate_limit')]
    def get(self, pstring):
        # The following position strings are supported
        # 1. 05 23 34.6 -69 45 22:0 6 (or 05h23m34.6s -69d45m22s:0m6s)
        # 2. 05 23 34.6 -69 45 22:0.166666 (or 05h23m34.6s -69d45m22s:0.166666)
        # 3. 80.89416667 -69.75611111:0.166666
        stime = time.time()
        # If we're given a string with qualifiers ('h', etc), convert to one without
        current_app.logger.info('Attempting SIMBAD position search: %s'%pstring)
        try:
            RA, DEC, radius = parse_position_string(pstring)
        except Exception, err:
            current_app.logger.error('Position string could not be parsed: %s' % pstring)
            return {'Error': 'Unable to get results!',
                    'Error Info': 'Invalid position string: %s'%pstring}, 200
        try:
            result = do_position_query(RA, DEC, radius)
        except timeout_decorator.timeout_decorator.TimeoutError:
            current_app.logger.error('Position query %s timed out' % pstring)
            return {'Error': 'Unable to get results!',
                    'Error Info': 'Position query timed out'}, 200
        return result

class QuerySearch(Resource):

    """Given a Solr query with object names, return a Solr query with SIMBAD identifiers"""
    scopes = []
    rate_limit = [1000, 60 * 60 * 24]
    decorators = [advertise('scopes', 'rate_limit')]

    def post(self):
        stime = time.time()
        # Get the supplied list of identifiers
        identifiers = []
        query = None
        itype = None
        name2id = {}
        try:
            query = request.json['query']
            input_type = 'query'
        except:
            current_app.logger.error('No query was specified for SIMBAD object search')
            return {"Error": "Unable to get results!",
                    "Error Info": "No identifiers/objects found in POST body"}, 200
        # If we get the request from BBB, the value of 'query' is actually an array
        if isinstance(query, list):
            try:
                solr_query = query[0]
            except:
                solr_query = ''
        else:
            solr_query = query
        current_app.logger.info('Received SIMBAD object query: %s'%solr_query)
        new_query = solr_query.replace('object:','simbid:')
        # If we receive a (Solr) query string, we need to parse out the object names
        try:
            identifiers = get_objects_from_query_string(solr_query)
        except:
            current_app.logger.error('Parsing the identifiers out of the query string blew up!')
            return {"Error": "Unable to get results!",
                    "Error Info": "No objects found in query string"}, 200
        identifiers = [iden for iden in identifiers if iden.lower() not in ['object',':']]
        # How many object names did we fid?
        id_num = len(identifiers)
        # Keep a list with the object names we found
        identifiers_orig = identifiers
        # If we did not find any object names, there is nothing to do!
        if id_num == 0:
            return {"Error": "Unable to get results!",
                    "Error Info": "No identifiers/objects found in POST body"}, 200
        # Source to query 
        source = 'simbad'
        if source in ['simbad','all'] and len(identifiers) > 0:
            if identifiers:
                for ident in identifiers:
                    result = get_simbad_data([ident], 'objects')
                    if 'Error' in result:
                        # An error was returned!
                        current_app.logger.error('Failed to find data for SIMBAD %s query!'%input_type)
                        return result
                    try:  
                        SIMBADid =[e.get('id',0) for e in result['data'].values()][0]
                    except:
                        SIMBADid = '0'
                    name2id[ident] = SIMBADid 
                for oname in identifiers:
                    try:
                        SIMBADid = name2id.get(oname)
                    except:
                        SIMBADid = '0'
                    new_query = new_query.replace(oname, SIMBADid)
                return {"query": new_query}
            else:
                # This should never happen
                current_app.logger.error('No data found, even though we had %s! Should never happen!'%input_type)
                result = {
                    "Error": "Failed to find data for SIMBAD %s query!"%input_type,
                    "Error Info": "No results found, where results were expected! Needs attention!"
                    }
                return result
